### PR TITLE
Rush publishing: hotfix versioning support.

### PR DESCRIPTION
**Pull request by dzearing**
When a dependency version range is bumped, we should be appending `-0` to allow for hotfixes to be consumed properly. This change updates the versioning logic to do so. Please see the unit test updates to get a better idea.

Here's the scenario:

utilities is v1.1.1
consumingLibrary v2.0.0 depends on utilities >=1.1.1 <2.0.0
consumingApp depends on consumingLibrary >=2.0.0 <3.0.0

If the app needs a hotfix in utilities, we publish 1.1.1-hotfix1. In the app, we add a dependency to utilities 1.1.1-hotfix1.

But now, because consumingLibrary has a 1.1.1 dependency that doesn't include the -0, it will not satisfy its semver and there will be 2 copies of utilities.

With this change, by enforcing the -0 in auto bumped dependencies, we can ensure that apps can consume hotfixes more easily without the duplicate dependency issue.

To experiment with semver satisfy, you can npm install `semver` package, and then run node in the folder. Then try a few things on the node command line:

```bash
> require('semver').satisfies('1.0.0-hotfix1', '>=1.0.0 <2.0.0')
false
> require('semver').satisfies('1.0.0-hotfix1', '^1.0.0')
false
> require('semver').satisfies('1.0.0-hotfix1', '^1.0.0-0')
true
> require('semver').satisfies('1.0.0-hotfix1', '>=1.0.0-0 <2.0.0-0')
true
>
```

### CODE REVIEW

**Comment by dzearing** [comment]
Note, I guess no change file is required because `shouldPublish` is set to false on rush.

File: apps/rush/src/utilities/PublishUtilities.ts [hunk 1]
```diff
@@ -139,7 +139,7 @@ export default class PublishUtilities {
   }
 
   public static isRangeDependency(version: string): boolean {
-    const LOOSE_PKG_REGEX: RegExp = />=?(?:\d+\.){2}\d+\s+<(?:\d+\.){2}\d+/;
+    const LOOSE_PKG_REGEX: RegExp = />=[\d\.\-]+\s<[\d\.\-]+/;
 
     return LOOSE_PKG_REGEX.test(version);
   }
```
Comment (nickpape-msft) [inline apps/rush/src/utilities/PublishUtilities.ts]: why aren't we using semver.validRange() ?

**Comment by nickpape-msft** [comment]
:shipit:

File: apps/rush/src/utilities/PublishUtilities.ts (hunk 1 above)
Comment (pgonzal) [inline apps/rush/src/utilities/PublishUtilities.ts]: +1

**Comment by qz2017** [comment]
@dzearing, when you publish a hotfix, can you publish 1.1.2 instead of 1.1.1-hotfix.1? Dash something in npm is treated as prerelease.

**Comment by pgonzal** [comment]
From http://semver.org/:

> 11. Precedence refers to how versions are compared to each other when ordered. Precedence MUST be calculated by separating the version into major, minor, patch and pre-release identifiers in that order (Build metadata does not figure into precedence). Precedence is determined by the first difference when comparing each of these identifiers from left to right as follows: Major, minor, and patch versions are always compared numerically. Example: 1.0.0 < 2.0.0 < 2.1.0 < 2.1.1. When major, minor, and patch are equal, a pre-release version has lower precedence than a normal version. Example: 1.0.0-alpha < 1.0.0. Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows: identifiers consisting of only digits are compared numerically and identifiers with letters or hyphens are compared lexically in ASCII sort order. Numeric identifiers always have lower precedence than non-numeric identifiers. A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal. Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.'

**Comment by dzearing** [comment]
@pgonzal The change is not about *selecting* the right dependencies, it is about *satisfying* semver.

If my library has a dependency on `"foo": "^1.0.0"`, then my app says to use `"foo": "1.0.0-hf1"`, the semver will not be satisfied and there will be 2 copies of "foo" in node_modules! This is what we're avoiding.

**Comment by dzearing** [comment]
@qz2017 Usually we can, but not always. If a partner is locked to a specific version for their ship build and need one specific fix, they do not want to pull in additional changes if necessary, hence creating a hotfix build. We may have already published patch N+1, N+2, etc beyond the one they're locked to. Just consuming a new patch means they'd need to go to N+3 which could include changes they didn't need.

So it is important that range dependencies are able to receive hotfix builds for any dependency.

**Comment by dzearing** [comment]
Another possibility is that we avoid using dashes for hotfix builds and instead use pluses. So the idea is this:

App needs to consume a hotfix of utilities 1.1.1
Control library uses utilities ^1.1.1 (and the app is shrinkwrapped to 1.1.1)

We release "1.1.1+hotfix1"

App updates shrinkwrap to consume 1.1.1+hotfix1
Control library semver is satisfied by this, so no dupe utilities

I'm just not 100% sure if this is problematic.

Re-reading the documentation about dashes, it is considered pre-release which isn't quite what we're looking for here, and I agree using the -0 in the dependencies works around the issue. If plus works, perhaps we can just use that.




```bash

**Comment by pgonzal** [comment]
@qz2017 is back, and pretty much owns the "rush publish" roadmap right now, so I will defer to her on this one.  :-)

@dzearing

**Comment by dzearing** [comment]
The more i think about it, I think what should happen is:

We should preserve the format that the package.json uses at all costs.

If they use ^x.x.x, we should preserve it and bump using ^x.x.x
If they use ^x.x.x-0, we should preserve that too

Currently, it will break if you use -0 and revert it to a locked version which is a bug.

If we agree on this goal, I can change this logic to not always append -0, but only append it if the original range includes it.

**Comment by pgonzal** [comment]
Wow, hello to @cliffkoh !  :-P

**Comment by qz2017** [comment]
@dzearing, I am covering some of that in my other pullrequest. Can you hold on until I merge that one in?

**Pull request closed without merging**